winsup/cygwin/include/cygwin/exit_process.h

```diff
@@ -0,0 +1,281 @@
+
+  HANDLE h[2];
+  SECURITY_ATTRIBUTES attrs;
+  attrs.nLength = sizeof(attrs);
+  attrs.bInheritHandle = TRUE;
+  attrs.lpSecurityDescriptor = NULL;
+  if (!CreatePipe(&h[0], &h[1], &attrs, 0))
+    return NULL;
+
+  STARTUPINFOW si = {};
+  PROCESS_INFORMATION pi;
+  size_t len = wcslen (wbuf) + wcslen(function_name) + 3 /* exit code */ + 1 /* space */ + 10 /* process ID, i.e. DWORD */ + 1 /* NUL */;
+  WCHAR cmd[len + 1];
+  WCHAR title[] = L"cygwin-console-helper";
+  LPDWORD process_exit;
+
+  swprintf (cmd, len + 1, L"%S %S %d %ld", wbuf, function_name, exit_code, pid);
+
+  si.cb = sizeof (si);
+  si.dwFlags = STARTF_USESHOWWINDOW | STARTF_USESTDHANDLES;
+  si.wShowWindow = SW_HIDE;
+  si.lpTitle = title;
+  si.hStdInput = si.hStdError = si.hStdOutput = INVALID_HANDLE_VALUE;
+
+  /* Create a new hidden process. */
+  BOOL x = CreateProcessW (NULL, cmd, NULL, NULL, TRUE,
+			   CREATE_NO_WINDOW | CREATE_NEW_PROCESS_GROUP,
+			   NULL, NULL, &si, &pi);
+  CloseHandle(h[1]);
+  if (!x)
+    {
+      CloseHandle(h[0]);
+      return NULL;
+    } else {
+    /* Wait for the process to complete for 10 seconds */
+      WaitForSingleObject(pi.hProcess, 10000 );
+    }
+
+  if (!GetExitCodeProcess(pi.hProcess, process_exit))
+    return NULL;
+
+  CloseHandle(pi.hThread);
+  CloseHandle(pi.hProcess);
+  CloseHandle(h[0]);
+
+  return process_exit;
+}
+
+static int current_is_wow = -1;
+static int is_32_bit_os = -1;
+
+static BOOL get_wow(HANDLE process, BOOL& is_wow, int& bitness)
+{
+  if (is_32_bit_os == -1)
+    {
+      SYSTEM_INFO info;
+
+      GetNativeSystemInfo(&info);
+      if (info.wProcessorArchitecture == 0)
+	is_32_bit_os = 1;
+      else if (info.wProcessorArchitecture == 9)
+	is_32_bit_os = 0;
+      else
+	is_32_bit_os = -2;
+    }
+
+  if (current_is_wow == -1 &&
+      !IsWow64Process (GetCurrentProcess (), &current_is_wow))
+    current_is_wow = -2;
+
+  if (is_32_bit_os == -2 || current_is_wow == -2)
+    return FALSE;
+
+  if (!IsWow64Process (process, &is_wow))
+    return FALSE;
+
+  bitness = is_32_bit_os || is_wow ? 32 : 64;
+  return TRUE;
+}
+
+/**
+ * Terminates the process corresponding to the process ID
+ *
+ * This way of terminating the processes is not gentle: the process gets
+ * no chance of cleaning up after itself (closing file handles, removing
+ * .lock files, terminating spawned processes (if any), etc).
+ */
+static int
+exit_one_process(HANDLE process, int exit_code, DWORD pid)
+{
+  LPTHREAD_START_ROUTINE address = NULL;
+  int signo = exit_code & 0x7f;
+  BOOL is_wow;
+  int bitness;
+  switch (signo)
+  {
+  case SIGINT:
+  case SIGQUIT:
+    if (get_wow(process, is_wow, bitness) &&
+        0 /* exit code 0 = success */ == kill_via_console_helper(bitness, L"CtrlRoutine",
+                                                                 signo == SIGINT ? CTRL_C_EVENT : CTRL_BREAK_EVENT, pid))
+      return 0;
+    /* fall-through */
+  case SIGTERM:
+    if (get_wow(process, is_wow, bitness) &&
```

Comment:
We probably should fold `get_wow()` into `kill_via_console_helper()`: this would also make the code a bit more thread-safe, in case we want to signal two processes concurrently (a `static int is_32_bit_process` is a single instance per process).

Reply:
Done https://github.com/msys2/msys2-runtime/pull/31/commits/e83163802298b9bf256ac437a8c325a28c1625b5